### PR TITLE
Sparql optimization

### PR DESCRIPTION
**Pull request by PonteIneptique**
I implemented in this PR a SleepyCat backend that is much more optimized than the SqlAlchemy one.

### CODE REVIEW

File: capitains_nautilus/cts/resolver.py [hunk 1]
```diff
@@ -464,4 +432,84 @@ def clear(self):
...
+            logger=logger,
+            cache=cache,
+            dispatcher=dispatcher
+        )
+
+        for exception in exceptions:
+            self.logger.warning(exception)
+
+
+class SleepyCatCTSResolver(_SparqlSharedResolver):
```
Comment (sonofmun) [inline capitains_nautilus/cts/resolver.py]: Probably should be `SleepyCatCtsResolver`

File: capitains_nautilus/cts/resolver.py [hunk 2]
```diff
@@ -464,4 +432,84 @@ def clear(self):
...
+            inventory_collection = type(self).CLASSES["inventory_collection"](identifier="defaultTic")
+            default_tiname = "/default"
+            ti = type(self).CLASSES["inventory"](default_tiname)
+            ti.parent = inventory_collection
+            if ti.get_label(lang="eng") is None:
+                ti.set_label("Default collection", "eng")
+            dispatcher = CollectionDispatcher(inventory_collection,
+                                              default_inventory_name=default_tiname)
+
+        super(SleepyCatCTSResolver, self).__init__(
```
Comment (sonofmun) [inline capitains_nautilus/cts/resolver.py]: Probably should be `SleepyCatCtsResolver`

File: capitains_nautilus/cts/resolver.py [hunk 3]
```diff
@@ -464,4 +432,84 @@ def clear(self):
...
+            logger=logger,
+            cache=cache,
+            dispatcher=dispatcher
+        )
+
+        for exception in exceptions:
+            self.logger.warning(exception)
+
+    def clear(self):
+        super(SleepyCatCTSResolver, self).clear()
```
Comment (sonofmun) [inline capitains_nautilus/cts/resolver.py]: Probably should be `SleepyCatCtsResolver`

File: setup.py [hunk 4]
```diff
@@ -15,8 +15,8 @@
     "Flask>=0.12",
     "Werkzeug>=0.11.3",
     "redis>=2.10.5",
-    "Flask-Caching==1.2.0",
-    "rdflib-sqlalchemy==0.3.8"
+    "Flask-Caching>=1.4.0,<2.0.0",
+    "rdflib-sqlalchemy==0.3.8",
```
Comment (sonofmun) [inline setup.py]: Is there a trailing comma here?

File: tests/cts/test_resolver_cache.py [hunk 5]
```diff
@@ -96,3 +98,32 @@ def x(*k, **kw):
...
+            clear_graph(sleepy_cat_address)
+        except:
+            print("Unable to clear graph")
+
+        output = call([python, "./tests/cts/scripts/run_cache_sparql_sleepy_cat.py"], cwd=cwd)
+        if output != 0:
+            raise Exception("Creating cache failed")
+
+        self.cache = FileSystemCache(subprocess_cache_dir)
+        self.resolver = SleepyCatCTSResolver(
```
Comment (sonofmun) [inline tests/cts/test_resolver_cache.py]: Probably should be `SleepyCatCtsResolver`

File: tests/cts/test_sparql_resolver/__init__.py [hunk 6]
```diff
@@ -0,0 +1,108 @@
+from tests.cts.test_resolver import TestXMLFolderResolverBehindTheScene, TextXMLFolderResolver, TextXMLFolderResolverDispatcher
+from capitains_nautilus.cts.resolver import SparqlAlchemyNautilusCTSResolver, SleepyCatCTSResolver
```
Comment (sonofmun) [inline tests/cts/test_sparql_resolver/__init__.py]: Probably should be `SleepyCatCtsResolver`

File: tests/cts/test_sparql_resolver/test_sleepy_cat.py [hunk 7]
```diff
@@ -0,0 +1,119 @@
+from tests.cts.test_resolver import TestXMLFolderResolverBehindTheScene, TextXMLFolderResolver, TextXMLFolderResolverDispatcher
+from capitains_nautilus.cts.resolver import SleepyCatCTSResolver
```
Comment (sonofmun) [inline tests/cts/test_sparql_resolver/test_sleepy_cat.py]: Probably should be `SleepyCatCtsResolver`

File: tests/cts/test_sparql_resolver/test_sleepy_cat.py [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+from capitains_nautilus.collections.sparql import clear_graph, \
+    generate_sleepy_cat_graph
+from MyCapytain.common.constants import set_graph
+from ..config import sleepy_cat_address
+from ...sparql_class import Sparql
+import shutil
+
+
+class _Parser(Sparql):
+    RESOLVER_CLASS = SleepyCatCTSResolver
```
Comment (sonofmun) [inline tests/cts/test_sparql_resolver/test_sleepy_cat.py]: Probably should be `SleepyCatCtsResolver`

File: tests/cts/test_sparql_resolver/__init__.py [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+from capitains_nautilus.collections.sparql import generate_alchemy_graph, clear_graph, \
+    generate_sleepy_cat_graph
+from MyCapytain.common.constants import set_graph
+from ..config import sqlite_address, sleepy_cat_address
+from ...sparql_class import Sparql
+
+
+
+class _Parser(Sparql):
+    RESOLVER_CLASS = SparqlAlchemyNautilusCTSResolver
```
Comment (sonofmun) [inline tests/cts/test_sparql_resolver/__init__.py]: Probably should be `SparqlAlchemyNautilusCtsResolver`

File: tests/cts/test_sparql_resolver/__init__.py [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+    def setUp(self):
+        self.generated_graphs = 0
+        self.gen_graph()
+        self.TEST_RUN += 1
+
+    def tearDown(self):
+        clear_graph(self.graph_identifier)
+
+
+class TextSparqlXMLFolderResolver(_Parser, TextXMLFolderResolver):
```
Comment (sonofmun) [inline tests/cts/test_sparql_resolver/__init__.py]: The two occurrences of `XML` here should probably be `Xml` instead.

File: tests/cts/test_sparql_resolver/__init__.py [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+    def setUp(self):
+        self.graph, self.graph_identifier, self.store_uri = generate_alchemy_graph(alchemy_uri=sqlite_address)
+        self.resolver = self.RESOLVER_CLASS(["./tests/testing_data/latinLit2"], graph=self.graph)
+        self.resolver.parse()
+
+    def tearDown(self):
+        clear_graph(self.graph_identifier)
+
+
+class TestSparqlXMLFolderResolverBehindTheScene(_Parser, TestXMLFolderResolverBehindTheScene):
```
Comment (sonofmun) [inline tests/cts/test_sparql_resolver/__init__.py]: The two occurrences of `XML` here should probably be `Xml` instead.

File: tests/cts/test_sparql_resolver/test_sleepy_cat.py [hunk 12]
```diff
@@ -0,0 +1,119 @@
+from tests.cts.test_resolver import TestXMLFolderResolverBehindTheScene, TextXMLFolderResolver, TextXMLFolderResolverDispatcher
```
Comment (sonofmun) [inline tests/cts/test_sparql_resolver/test_sleepy_cat.py]: The occurrences of `XML` here should probably be `Xml` instead.

File: tests/cts/test_sparql_resolver/test_sleepy_cat.py [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+
+class _Parser(Sparql):
+    RESOLVER_CLASS = SleepyCatCTSResolver
+
+
+def remove_sleepy_cat(address):
+    shutil.rmtree(address, ignore_errors=True)
+
+
+class TestSparqlSleepyCatBasedResolverDispatcher(_Parser, TextXMLFolderResolverDispatcher):
```
Comment (sonofmun) [inline tests/cts/test_sparql_resolver/test_sleepy_cat.py]: `TextXMLFolderResolverDispatcher` here should probably be `TextXmlFolderResolverDispatcher` instead.

File: tests/cts/test_sparql_resolver/test_sleepy_cat.py [hunk 14]
```diff
@@ -0,0 +1,119 @@
...
+        remove_sleepy_cat(sleepy_cat_address)
+        self.generated_graphs = 0
+        self.gen_graph()
+        self.TEST_RUN += 1
+
+    def tearDown(self):
+        clear_graph(self.graph_identifier)
+
+
+class TextSleepyCatSparqlXMLFolderResolver(_Parser, TextXMLFolderResolver):
```
Comment (sonofmun) [inline tests/cts/test_sparql_resolver/test_sleepy_cat.py]: The two occurrences of `XML` here should probably be `Xml`

File: tests/cts/test_sparql_resolver/test_sleepy_cat.py [hunk 15]
```diff
@@ -0,0 +1,119 @@
...
+        )
+        self.resolver = self.RESOLVER_CLASS(["./tests/testing_data/latinLit2"], graph=self.graph)
+        self.resolver.parse()
+
+    def tearDown(self):
+        remove_sleepy_cat(sleepy_cat_address)
+        self.resolver.clear()
+
+
+class TestSleepyCatSparqlXMLFolderResolverBehindTheScene(_Parser, TestXMLFolderResolverBehindTheScene):
```
Comment (sonofmun) [inline tests/cts/test_sparql_resolver/test_sleepy_cat.py]: The two occurrences of `XML` here should probably be `Xml`

File: tests/cts/test_sparql_resolver/__init__.py [hunk 16]
```diff
@@ -0,0 +1,108 @@
...
+from ..config import sqlite_address, sleepy_cat_address
+from ...sparql_class import Sparql
+
+
+
+class _Parser(Sparql):
+    RESOLVER_CLASS = SparqlAlchemyNautilusCTSResolver
+
+
+class TestSparqlBasedResolverDispatcher(_Parser, TextXMLFolderResolverDispatcher):
```
Comment (sonofmun) [inline tests/cts/test_sparql_resolver/__init__.py]: `TextXMLFolderResolverDispatcher` should probably be `TextXmlFolderResolverDispatcher`

File: tests/cts/test_sparql_resolver/__init__.py [hunk 17]
```diff
@@ -0,0 +1,108 @@
+from tests.cts.test_resolver import TestXMLFolderResolverBehindTheScene, TextXMLFolderResolver, TextXMLFolderResolverDispatcher
```
Comment (sonofmun) [inline tests/cts/test_sparql_resolver/__init__.py]: The occurrences of `XML` here should probably be `Xml`

File: tests/cts/test_sparql_resolver/test_sqlalchemy.py [hunk 18]
```diff
@@ -2,8 +2,9 @@
 from capitains_nautilus.cts.resolver import SparqlAlchemyNautilusCTSResolver
```
Comment (sonofmun) [inline tests/cts/test_sparql_resolver/test_sqlalchemy.py]: `SparqlAlchemyNautilusCTSResolver` should probably be `SparqlAlchemyNautilusCtsResolver`

File: capitains_nautilus/cts/collections.py [hunk 19]
```diff
@@ -14,6 +14,7 @@
 from capitains_nautilus.collections.sparql import SparqlNavigatedCollection
 from rdflib import BNode, Literal, RDF, URIRef
 from capitains_nautilus.errors import UnknownCollection
+from capitains_nautilus.utils.performances import cached_property, STORE
 
 
 class CTSSparqlNavigatedCollection(PrototypeCtsCollection, SparqlNavigatedCollection):
```
Comment (sonofmun) [inline capitains_nautilus/cts/collections.py]: `CTSSparqlNavigatedCollection` should probably be `CtsSparqlNavigatedCollection`

File: capitains_nautilus/cts/collections.py [hunk 20]
```diff
@@ -168,8 +157,13 @@ class SparqlXmlCtsEditionMetadata(SparqlXmlCtsTextMetadata, XmlCtsEditionMetadat
 
 
 class SparqlXmlCtsWorkMetadata(CTSSparqlNavigatedCollection, XmlCtsWorkMetadata):
```
Comment (sonofmun) [inline capitains_nautilus/cts/collections.py]: `CTSSparqlNavigatedCollection` should probably be `CtsSparqlNavigatedCollection`

File: capitains_nautilus/cts/collections.py [hunk 21]
```diff
@@ -205,12 +199,14 @@ def decide_class(self, key):
 
 
 class SparqlXmlCtsTextgroupMetadata(CTSSparqlNavigatedCollection, XmlCtsTextgroupMetadata):
```
Comment (sonofmun) [inline capitains_nautilus/cts/collections.py]: `CTSSparqlNavigatedCollection` should probably be `CtsSparqlNavigatedCollection`

File: capitains_nautilus/cts/resolver.py [hunk 22]
```diff
@@ -25,7 +27,7 @@
     SparqlTextInventoryCollection
 )
 from capitains_nautilus.resolver_prototype import NautilusPrototypeResolver
-
+from capitains_nautilus.utils.dev import DevPool as ThreadPool
 
 class __BaseNautilusCTSResolver__(CtsCapitainsLocalResolver, NautilusPrototypeResolver):
```
Comment (sonofmun) [inline capitains_nautilus/cts/resolver.py]: `__BaseNautilusCTSResolver__` should probably be `__BaseNautilusCtsResolver__`

File: capitains_nautilus/cts/resolver.py [hunk 23]
```diff
@@ -312,7 +314,7 @@ def clear(self):
         return self.__cache__.clear()
 
 
-class SparqlAlchemyNautilusCTSResolver(__BaseNautilusCTSResolver__):
+class _SparqlSharedResolver(__BaseNautilusCTSResolver__):
```
Comment (sonofmun) [inline capitains_nautilus/cts/resolver.py]: `__BaseNautilusCTSResolver__` should probably be `__BaseNautilusCtsResolver__`

File: capitains_nautilus/cts/resolver.py [hunk 24]
```diff
@@ -464,4 +432,84 @@ def clear(self):
         """ Deletes the database
         """
         clear_graph(self.graph_identifier)
-        super(SparqlAlchemyNautilusCTSResolver, self).clear()
+        super(_SparqlSharedResolver, self).clear()
+
+
+class SparqlAlchemyNautilusCTSResolver(_SparqlSharedResolver):
```
Comment (sonofmun) [inline capitains_nautilus/cts/resolver.py]: `SparqlAlchemyNautilusCTSResolver` should probably be `SparqlAlchemyNautilusCtsResolver`

File: capitains_nautilus/cts/resolver.py [hunk 25]
```diff
@@ -464,4 +432,84 @@ def clear(self):
...
+            inventory_collection = type(self).CLASSES["inventory_collection"](identifier="defaultTic")
+            default_tiname = "/default"
+            ti = type(self).CLASSES["inventory"](default_tiname)
+            ti.parent = inventory_collection
+            if ti.get_label(lang="eng") is None:
+                ti.set_label("Default collection", "eng")
+            dispatcher = CollectionDispatcher(inventory_collection,
+                                              default_inventory_name=default_tiname)
+
+        super(SparqlAlchemyNautilusCTSResolver, self).__init__(
```
Comment (sonofmun) [inline capitains_nautilus/cts/resolver.py]: `SparqlAlchemyNautilusCTSResolver` should probably be `SparqlAlchemyNautilusCtsResolver`

File: tests/cts/scripts/run_cache.py [hunk 26]
```diff
@@ -2,7 +2,7 @@
 import sys
 
 sys.path.append(
-    os.path.join(os.path.dirname(__file__), "..", "..")
+    os.path.join(os.path.dirname(__file__), "..", "..", "..")
 )
 
 from capitains_nautilus.cts.resolver import NautilusCTSResolver
```
Comment (sonofmun) [inline tests/cts/scripts/run_cache.py]: `NautilusCTSResolver` should probably be `NautilusCtsResolver`

File: tests/cts/scripts/run_cache_sparql.py [hunk 27]
```diff
@@ -2,7 +2,7 @@
 import sys
 
 sys.path.append(
-    os.path.join(os.path.dirname(__file__), "..", "..")
+    os.path.join(os.path.dirname(__file__), "..", "..", "..")
 )
 
 from capitains_nautilus.cts.resolver import SparqlAlchemyNautilusCTSResolver
```
Comment (sonofmun) [inline tests/cts/scripts/run_cache_sparql.py]: `SparqlAlchemyNautilusCTSResolver` should probably be `SparqlAlchemyNautilusCtsResolver`

File: tests/cts/scripts/run_cache_sparql_sleepy_cat.py [hunk 28]
```diff
@@ -0,0 +1,18 @@
+import os
+import sys
+
+sys.path.append(
+    os.path.join(os.path.dirname(__file__), "..", "..", "..")
+)
+
+from capitains_nautilus.cts.resolver import SleepyCatCTSResolver
```
Comment (sonofmun) [inline tests/cts/scripts/run_cache_sparql_sleepy_cat.py]: `SleepyCatCTSResolver` should probably be `SleepyCatCtsResolver`

File: tests/cts/scripts/run_cache_sparql_sleepy_cat.py [hunk 29]
```diff
@@ -0,0 +1,18 @@
...
+from werkzeug.contrib.cache import FileSystemCache
+from tests.cts.config import subprocess_repository, subprocess_cache_dir, sleepy_cat_address
+
+cache = FileSystemCache(subprocess_cache_dir)
+resolver = SleepyCatCTSResolver(
```
Comment (sonofmun) [inline tests/cts/scripts/run_cache_sparql_sleepy_cat.py]: `SleepyCatCTSResolver` should probably be `SleepyCatCtsResolver`

File: tests/cts/test_resolver_cache.py [hunk 30]
```diff
@@ -2,11 +2,13 @@
 from subprocess import call
 from sys import executable
 import os
-from capitains_nautilus.cts.resolver import NautilusCTSResolver, SparqlAlchemyNautilusCTSResolver
+from capitains_nautilus.cts.resolver import NautilusCTSResolver, \
```
Comment (sonofmun) [inline tests/cts/test_resolver_cache.py]: The three `CTS` occurrences in this import statement should probably be `Cts`

**Review by sonofmun — COMMENTED** [review]
I am certainly not the best person to review this PR since I don't have a good overview of the Nautilus code or of sparql. I have made some suggestions mostly for the naming of classes/functions. Do with them as you wish.

**Comment by PonteIneptique** [comment]
The PR has been corrected and I added some other good practices in there :)

**Pull request merged**